File: android/app/src/main/java/ai/thetahealth/mirobody/ui/chat/HistoryViewModel.kt
package ai.thetahealth.mirobody.ui.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ai.thetahealth.mirobody.data.chat.ChatRepository
import ai.thetahealth.mirobody.data.chat.dto.SessionSummary
import ai.thetahealth.mirobody.data.net.ErrorBus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class HistoryUiState(
    val items: List<SessionSummary> = emptyList(),
    val loading: Boolean = false,
    val error: String? = null,
)

class HistoryViewModel(
    private val repo: ChatRepository,
    private val errorBus: ErrorBus,
) : ViewModel() {
    private val _state = MutableStateFlow(HistoryUiState())
    val state: StateFlow<HistoryUiState> = _state.asStateFlow()

    fun refresh() {
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { repo.history() }
                .onSuccess { items -> _state.update { it.copy(loading = false, items = items) } }
                .onFailure { t ->
                    errorBus.emit(t)
                    _state.update {
                        it.copy(loading = false, error = t.message ?: "Failed to load history")
                    }
                }
        }
    }

    fun deleteHistory(sessionId: String) {
        if (sessionId.isBlank()) return
        // Optimistically drop the row; restore if the server call fails. The user is
        // informed of the failure via the app-wide Snackbar (ErrorBus).
        val previous = _state.value.items
        _state.update { it.copy(items = previous.filterNot { row -> row.sessionId == sessionId }) }
        viewModelScope.launch {
            runCatching { repo.deleteHistory(sessionId) }
                .onFailure { t ->
                    errorBus.emit(t)
                    _state.update { it.copy(items = previous) }
                }
        }
    }
}

File: ios/Mirobody/UI/Chat/HistoryViewModel.swift
import Foundation

/// Drives the history screen — mirrors `ui/chat/HistoryViewModel.kt`.
@MainActor
final class HistoryViewModel: ObservableObject {
    @Published private(set) var items: [SessionSummary] = []
    @Published private(set) var loading = false
    @Published private(set) var error: String?

    private let repo: ChatRepository
    private let errorBus: ErrorBus
    private let language: String

    init(repo: ChatRepository, errorBus: ErrorBus, language: String) {
        self.repo = repo
        self.errorBus = errorBus
        self.language = language
    }

    func refresh() {
        loading = true
        error = nil
        Task {
            do {
                items = try await repo.history()
                loading = false
            } catch {
                errorBus.emit(error)
                loading = false
                self.error = localizedMessage(error.toAppError(), language: language)
            }
        }
    }

    func deleteHistory(sessionId: String) {
        guard !sessionId.isEmpty else { return }
        // Optimistically drop the row; restore if the server call fails. The failure
        // is surfaced via the app-wide toast (ErrorBus).
        let previous = items
        items.removeAll { $0.sessionId == sessionId }
        Task {
            do {
                try await repo.deleteHistory(sessionId: sessionId)
            } catch {
                errorBus.emit(error)
                items = previous
            }
        }
    }
}
